File: tasks/native-runtime/tool_catalog_parity_probe.py
"""Tool-catalog parity probe, Python side.

Covers `available_tool_definitions` / `tool_parameter_schemas` / `schema_for_tool` /
`agent_tool_definitions` from `infra/tool_runtime/tools.py`.

Unlike the other probes in this directory, this one **imports the module directly**
rather than AST-extracting the functions. Extraction would mean chasing a transitive
dependency chain (`available_tool_definitions` -> `browser_tool_definitions` /
`additional_tool_definitions` -> `mindmap_node_schema` / ...) and re-implementing the
module's own scaffolding; the module imports cleanly, so running the real thing is both
simpler and more faithful.

The strongest case is `asset::json`: the embedded Rust asset must be byte-identical to
`json.dumps(available_tool_definitions(), ensure_ascii=False, indent=2)`. That is what
proves the committed 40 KB file is still the oracle's own rendering rather than a
hand-edited copy.

Usage::

    python tasks/native-runtime/tool_catalog_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example tool_catalog_parity_probe > ../rust.json
    diff <(tr -d '\\r' < python.json) <(tr -d '\\r' < rust.json)
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))

SCHEMA_CASES = [
    "web_search",
    "create_reminder",
    "generate_chart",
    "read_file_chunk",
    "browser_open_url",
    "  web_search  ",
    "no_such_tool",
    "",
]

INDEX_SAMPLE = ["web_search", "create_reminder", "generate_chart", "read_file_chunk"]


def main() -> int:
    from deepseek_infra.infra.tool_runtime.tools import (
        agent_tool_definitions,
        available_tool_definitions,
        schema_for_tool,
        tool_parameter_schemas,
    )

    tools = available_tool_definitions()
    out: dict = {}

    # The asset's exact bytes, as the oracle renders them.
    out["asset::json"] = json.dumps(tools, ensure_ascii=False, indent=2)

    # Declaration order is load-bearing: callers index into it.
    out["definitions::names"] = [
        str(tool.get("function", {}).get("name") or "") for tool in tools
    ]
    out["definitions::count"] = len(tools)

    index = tool_parameter_schemas()
    out["schemas::count"] = len(index)
    out["schemas::names"] = sorted(index)
    for name in INDEX_SAMPLE:
        out[f"schema::{name}"] = index.get(name)

    for case in SCHEMA_CASES:
        out[f"schema-for::{json.dumps(case, ensure_ascii=False)}"] = schema_for_tool(case)

    # With no external MCP profiles registered this is the local catalog unchanged, which
    # is the arm the Rust side passes `None` for.
    out["agent::names"] = [
        str(tool.get("function", {}).get("name") or "") for tool in agent_tool_definitions()
    ]

    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/tool_catalog_parity_probe.rs
//! Tool-catalog parity probe, Rust side.
//!
//! Replays the same corpus as `tasks/native-runtime/tool_catalog_parity_probe.py`
//! through `deepseek_policy::tool_catalog`.
//!
//! Usage::
//!
//!     python tasks/native-runtime/tool_catalog_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example tool_catalog_parity_probe > ../rust.json
//!     diff <(tr -d '\r' < python.json) <(tr -d '\r' < rust.json)

use deepseek_policy::tool_catalog::{
    agent_tool_definitions, available_tool_definitions, catalog_json, schema_for_tool,
    tool_parameter_schemas,
};
use serde_json::{Map, Value, json};

fn schema_cases() -> Vec<&'static str> {
    vec![
        "web_search",
        "create_reminder",
        "generate_chart",
        "read_file_chunk",
        "browser_open_url",
        "  web_search  ",
        "no_such_tool",
        "",
    ]
}

fn index_sample() -> [&'static str; 4] {
    [
        "web_search",
        "create_reminder",
        "generate_chart",
        "read_file_chunk",
    ]
}

fn main() {
    let mut out = Map::new();

    // The asset's exact bytes, as the oracle renders them. Byte-identity here is what
    // proves the committed JSON is still `json.dumps(..., ensure_ascii=False, indent=2)`.
    out.insert("asset::json".to_string(), json!(catalog_json()));

    let names: Vec<Value> = available_tool_definitions()
        .iter()
        .map(|tool| tool["function"]["name"].clone())
        .collect();
    out.insert("definitions::names".to_string(), json!(names));
    out.insert(
        "definitions::count".to_string(),
        json!(available_tool_definitions().len()),
    );

    let index = tool_parameter_schemas();
    out.insert("schemas::count".to_string(), json!(index.len()));
    out.insert(
        "schemas::names".to_string(),
        json!(index.keys().cloned().collect::<Vec<String>>()),
    );
    for name in index_sample() {
        out.insert(
            format!("schema::{name}"),
            index.get(name).cloned().unwrap_or(Value::Null),
        );
    }

    for case in schema_cases() {
        // The key mirrors Python's `json.dumps(case, ensure_ascii=False)`.
        out.insert(
            format!("schema-for::{}", json!(case)),
            schema_for_tool(case, None).unwrap_or(Value::Null),
        );
    }

    // No external MCP profiles: the local catalog unchanged, which is the arm the
    // oracle takes when the bridge contributes nothing.
    let agent: Vec<Value> = agent_tool_definitions(None)
        .iter()
        .map(|tool| tool["function"]["name"].clone())
        .collect();
    out.insert("agent::names".to_string(), json!(agent));

    let mut encoded =
        serde_json::to_string_pretty(&Value::Object(out)).expect("serialize probe output");
    encoded.push('\n');
    print!("{encoded}");
}
